griptape/loaders/csv_loader.py

```diff
@@ -13,27 +15,34 @@
 class CsvLoader(BaseLoader):
     embedding_driver: Optional[BaseEmbeddingDriver] = field(default=None, kw_only=True)
     delimiter: str = field(default=",", kw_only=True)
+    encoding: str = field(default="utf-8", kw_only=True)
 
-    def load(self, source: str, *args, **kwargs) -> list[CsvRowArtifact]:
-        return self._load_file(source)
-
-    def load_collection(self, sources: list[str], *args, **kwargs) -> dict[str, list[CsvRowArtifact]]:
-        return utils.execute_futures_dict(
-            {utils.str_to_hash(source): self.futures_executor.submit(self._load_file, source) for source in sources}
-        )
-
-    def _load_file(self, filename: str) -> list[CsvRowArtifact]:
+    def load(self, source: bytes | str, *args, **kwargs) -> ErrorArtifact | list[CsvRowArtifact]:
         artifacts = []
 
-        with open(filename, encoding="utf-8") as csv_file:
-            reader = csv.DictReader(csv_file, delimiter=self.delimiter)
-            chunks = [CsvRowArtifact(row) for row in reader]
+        if isinstance(source, bytes):
+            try:
+                source = source.decode(encoding=self.encoding)
+            except UnicodeDecodeError:
+                return ErrorArtifact(f"Failed to decode bytes to string using encoding: {self.encoding}")
+        elif isinstance(source, bytearray) or isinstance(source, memoryview):
+            return ErrorArtifact(f"Unsupported source type: {type(source)}")
```

Comment:
Wouldn't a type checker also prevent this case since we've got the type as `bytes | str`?

Reply:
The type checker is the one who decided this was necessary! I'm digging through the docs trying to find out why.

Reply:
Here is the source of the problem:

> ...the `typing` module docs state that type checkers should treat `bytearray` and `memoryview` as implicit subtypes of `bytes`.

From https://github.com/python/mypy/issues/12643#issuecomment-1105580582

(Side note, I noticed from that thread that I can apply slightly more concise syntax in the form of a tuple when passin the type argument - `isinstance(source, (bytearray, memoryview))`)